substrate/primitives/core/src/pop.rs

```diff
@@ -0,0 +1,123 @@
+
+use crate::crypto::{CryptoType, Pair};
+use sp_std::vec::Vec;
+
+/// Pair which is able to generate proof of possession.
+///
+/// This is implemented in different trait to provide default behavior.
+pub trait ProofOfPossessionGenerator: Pair
+where
+	Self::Public: CryptoType,
+{
+	/// Generate proof of possession.
+	///
+	/// The proof of possession generator is supposed to
+	/// produce a "signature" with unique hash context that should
+	/// never be used in other signatures. This proves that
+	/// the secret key is known to the prover. While prevent
+	/// malicious actors to trick an honest party to sign an
+	/// unpossessed public key resulting in a rogue key attack (See: Section 4.3 of
+	/// - Ristenpart, T., & Yilek, S. (2007). The power of proofs-of-possession: Securing multiparty
+	///   signatures against rogue-key attacks. In , Annual {{International Conference}} on the
+	///   {{Theory}} and {{Applications}} of {{Cryptographic Techniques} (pp. 228–245). : Springer.
+	#[cfg(feature = "full_crypto")]
+	fn generate_proof_of_possession(&mut self) -> Self::Signature;
+}
+
+/// Pair which is able to generate proof of possession.
+///
+/// While you don't need a keypair to verify a proof of possession (you only need a public key)
+/// we constrain on Pair to use the Public and Signature types associated to Pair.
+/// This is implemented in different trait (than Public Key) to provide default behavior.
+pub trait ProofOfPossessionVerifier: Pair
+where
+	Self::Public: CryptoType,
+{
+	/// Verify proof of possession.
+	///
+	/// The proof of possession verifier is supposed to to verify a signature with unique hash
+	/// context that is produced solely for this reason. This proves that that the secret key is
+	/// known to the prover.
+	fn verify_proof_of_possession(
+		proof_of_possession: &Self::Signature,
+		allegedly_possessesd_pubkey: &Self::Public,
+	) -> bool;
+}
+
+/// Marker trait to identify whether the scheme is not aggregatable.
+///
+/// Aggregatable schemes may change/optimize implementation parts such as Proof Of Possession
+/// or other specifics.
+///
+/// This is specifically because implementation of proof of possession for aggregatable schemes
+/// is security critical.
+///
+/// We would like to prevent aggregatable scheme from unknowingly generating signatures
+/// which aggregate to false albeit valid proof of possession aka rouge key attack.
+/// We ensure that by separating signing and generating pop at the API level.
+///
+/// Rouge key attack however is not immediately applicable to non-aggregatable scheme
+/// when even if an honest signing oracle is tricked to sign a rogue pop, it is not
+/// possible to aggregate it to generate a valid proof for a key the attack does not
+/// possess. Therefore we do not require non-aggregatable schemes to prevent PoP
+/// confirming signatures at API level
+pub trait NonAggregatable: Pair {
+	/// Default PoP statement.
+	fn pop_statement(pk: &impl crate::Public) -> Vec<u8> {
```

Comment:
```suggestion
	fn proof_of_possession_statement(pk: &impl crate::Public) -> Vec<u8> {
```

Reply:
We are using a well-adopted acronym used across the security industry (see  https://www.ietf.org/archive/id/draft-ietf-oauth-pop-architecture-08.txt for example).

 For example, I don't think we should rename all use of `json` in the code to `Javascript_object_notation`.

Reply:
Yeah and `pop` is also used everywhere to remove something from the end of a vec/array/list whatever. So, while reading this name I would first thing you remove some statement. Also we are writing here code and not security papers/specs :)